pydantic/main.py

```diff
@@ -57,6 +64,97 @@
 
 _object_setattr = _model_construction.object_setattr
 
+if sys.version_info >= (3, 10):
+
+    def _is_union(origin):
+        return origin is typing.Union or origin is types.UnionType
+
+else:
+
+    def _is_union(origin):
+        return origin is typing.Union
+
+
+def _handle_union(args, value):
+    # TODO: union_mode/discriminators?
+    for possible_type in args:
+        result = _recursive_model_construct(possible_type, value)
+        if result != value:
+            return result
+
+    # Otherwise, no union member matched
+    return value
+
+
+def _recursive_model_construct(annotation: type | None, value: Any):
+    # No annotation? Early exit
+    if annotation is None:
+        return value
+
+    # If value is already a BaseModel, just return that
+    # NOTE: if we see this, we assume that all of the children of the supplied BaseModel have been recursed through and
+    # converted to BaseModels, OR that the user manually provided this value and would like us to *not* recurse through
+    # its children and instead leave it "as-is". Depending on the circumstance, the user might want either behavior, so
+    # perhaps this could be toggled with another parameter/setting.
+    if isinstance(value, BaseModel):
+        return value
+
+    # Try to get a "parent" class if we're dealing with a structural type like list, dict, sequence, etc.
+    # Otherwise just treat the annotation itself as the parent
+    origin = get_origin(annotation)
```

Comment:
Haven't thought about Pydantic generics at all yet; determining the subclassing for regular generics is already painful enough. Thanks for the tip, though; I'll take a look at how much more complex it would be to include Pydantic ones.